File: bwl.Network.ClientServer/Remoting/RemoteAppClient.vb
Imports System.Linq
Imports System.Windows.Forms
Imports bwl.Framework

Public Class RemoteAppClient
    Private _settingsClients As New List(Of SettingsClient)
    Private _logsClients As New List(Of LogsClient)
    Private _autoUiClients As New List(Of AutoUiClient)
    Private _prefixes As New List(Of String)

    Public ReadOnly Property SettingsClient As SettingsClient
        Get
            Return _settingsClients.FirstOrDefault
        End Get
    End Property

    Public ReadOnly Property LogsClient As LogsClient
        Get
            Return _logsClients.FirstOrDefault
        End Get
    End Property

    Public ReadOnly Property AutoUiClient As AutoUiClient
        Get
            Return _autoUiClients.FirstOrDefault
        End Get
    End Property

    Public ReadOnly Property Prefix As String
        Get
            Return _prefixes.FirstOrDefault
        End Get
    End Property

    Public ReadOnly Property SettingsClients As SettingsClient()
        Get
            Return _settingsClients.ToArray()
        End Get
    End Property

    Public ReadOnly Property LogsClients As LogsClient()
        Get
            Return _logsClients.ToArray()
        End Get
    End Property

    Public ReadOnly Property AutoUiClients As AutoUiClient()
        Get
            Return _autoUiClients.ToArray()
        End Get
    End Property

    Public ReadOnly Property Prefixes As String()
        Get
            Return _prefixes.ToArray()
        End Get
    End Property

    Public ReadOnly Property MessageTransport As IMessageTransport

    Private _settingsForm As SettingsDialog
    Private _invokeFrom As Form
    Private _formTitle As String
    Private _createdForm As Form

    Public Sub New()
        Me.New("remote-app", "")
    End Sub

    Public Sub New(prefix As String, target As String)
        Me.New(New NetClient, prefix, target)
    End Sub

    Public Sub New(prefix As IEnumerable(Of String), target As String)
        Me.New(New NetClient, prefix, target)
    End Sub

    Public Sub New(netClient As IMessageTransport, prefix As String, target As String)
        Me.New(netClient, {prefix}, target)
    End Sub

    Public Sub New(netClient As IMessageTransport, prefix As IEnumerable(Of String), target As String)
        _MessageTransport = netClient
        For i = 0 To prefix.Count - 1
            _settingsClients.Add(New SettingsClient(_MessageTransport, prefix(i), target))
            _logsClients.Add(New LogsClient(_MessageTransport, prefix(i), target))
            _autoUiClients.Add(New AutoUiClient(_MessageTransport, prefix(i), target))
            _prefixes.Add(prefix(i))
        Next
    End Sub

    Public Sub Connect(address As String, Optional options As String = "")
        _MessageTransport.Open(address, options)
    End Sub

    Public Sub Dispose()
        If _MessageTransport.IsConnected Then _MessageTransport.Close()
        _MessageTransport = Nothing

        For Each sc In _settingsClients
            sc.Dispose()
            sc = Nothing
        Next
        _settingsClients.Clear()

        For Each lc In _logsClients
            lc.Dispose()
            lc = Nothing
        Next
        _logsClients.Clear()

        For Each ui In _autoUiClients
            ui.Dispose()
            ui = Nothing
        Next
        _autoUiClients.Clear()

        If _createdForm IsNot Nothing Then
            Try
                _createdForm.Close()
                _createdForm.Dispose()
            Catch ex As Exception
            End Try
            _createdForm = Nothing
        End If
    End Sub

    Public Function CreateAutoUiForm() As AutoUIForm
        Return CreateAutoUiForm(Prefix)
    End Function

    Public Function CreateAutoUiForm(prefix As String) As AutoUIForm
        If _createdForm IsNot Nothing Then
            Try
                _createdForm.Close()
                _createdForm.Dispose()
            Catch ex As Exception
            End Try
            _createdForm = Nothing
        End If
        For i = 0 To _prefixes.Count - 1
            If _prefixes(i) = prefix Then
                _createdForm = AutoUIForm.Create(_settingsClients(i), _logsClients(i), _autoUiClients(i))
                _createdForm.Text += " RemoteApp"
                Exit For
            End If
        Next
        Return _createdForm
    End Function

    Public Sub RunRemoteApp()
        Application.EnableVisualStyles()
        CreateAutoUiForm.Show()
        Application.Run()
    End Sub
End Class

File: ClientServer/ClientServer/Bwl.Network.ClientServer.Avalonia/Remoting/RemoteAppClient.cs
using System.Runtime.Versioning;
using Bwl.Framework.Avalonia;
using Bwl.Framework;
using Avalonia.Controls;
using Avalonia.Threading;

namespace Bwl.Network.ClientServer.Avalonia
{
    [SupportedOSPlatform("windows")]
    public partial class RemoteAppClient
    {
        private readonly List<SettingsClient> _settingsClients = new List<SettingsClient>();
        private readonly List<LogsClient> _logsClients = new List<LogsClient>();
        private readonly List<AutoUiClient> _autoUiClients = new List<AutoUiClient>();
        private readonly List<string> _prefixes = new List<string>();

        public SettingsClient SettingsClient
        {
            get
            {
                return _settingsClients.FirstOrDefault();
            }
        }

        public LogsClient LogsClient
        {
            get
            {
                return _logsClients.FirstOrDefault();
            }
        }

        public AutoUiClient AutoUiClient
        {
            get
            {
                return _autoUiClients.FirstOrDefault();
            }
        }

        public string Prefix
        {
            get
            {
                return _prefixes.FirstOrDefault();
            }
        }

        public SettingsClient[] SettingsClients
        {
            get
            {
                return _settingsClients.ToArray();
            }
        }

        public LogsClient[] LogsClients
        {
            get
            {
                return _logsClients.ToArray();
            }
        }

        public AutoUiClient[] AutoUiClients
        {
            get
            {
                return _autoUiClients.ToArray();
            }
        }

        public string[] Prefixes
        {
            get
            {
                return _prefixes.ToArray();
            }
        }

        public IMessageTransport MessageTransport { get; private set; }

        private SettingsDialog _settingsForm;
        private Window _invokeFrom;
        private string _formTitle;
        private IAutoUIForm _createdForm;

        public RemoteAppClient() : this("remote-app", "")
        {
        }

        public RemoteAppClient(string prefix, string target) : this(new NetClient(), prefix, target)
        {
        }

        public RemoteAppClient(IEnumerable<string> prefix, string target) : this(new NetClient(), prefix, target)
        {
        }

        public RemoteAppClient(IMessageTransport netClient, string prefix, string target) : this(netClient, new[] { prefix }, target)
        {
        }

        public RemoteAppClient(IMessageTransport netClient, IEnumerable<string> prefix, string target)
        {
            MessageTransport = netClient;
            for (int i = 0, loopTo = prefix.Count() - 1; i <= loopTo; i++)
            {
                _settingsClients.Add(new SettingsClient(MessageTransport, prefix.ElementAtOrDefault(i), target));
                _logsClients.Add(new LogsClient(MessageTransport, prefix.ElementAtOrDefault(i), target));
                _autoUiClients.Add(new AutoUiClient(MessageTransport, prefix.ElementAtOrDefault(i), target));
                _prefixes.Add(prefix.ElementAtOrDefault(i));
            }
        }

        public void Connect(string address, string options = "")
        {
            MessageTransport.Open(address, options);
        }

        public void Dispose()
        {
            if (MessageTransport.IsConnected)
                MessageTransport.Close();
            MessageTransport = null;

            foreach (var sc in _settingsClients)
            {
                sc.Dispose();
            }
            _settingsClients.Clear();

            foreach (var lc in _logsClients)
            {
                lc.Dispose();
            }
            _logsClients.Clear();

            foreach (var ui in _autoUiClients)
            {
                ui.Dispose();
            }
            _autoUiClients.Clear();

            if (_createdForm is not null)
            {
                try
                {
                    _createdForm.Close();
                    _createdForm.Dispose();
                }
                catch (Exception ex)
                {
                }
                _createdForm = null;
            }
        }

        public IAutoUIForm CreateAutoUiForm()
        {
            return CreateAutoUiForm(Prefix);
        }

        public IAutoUIForm CreateAutoUiForm(string prefix)
        {
            if (_createdForm is not null)
            {
                try
                {
                    _createdForm.Close();
                    _createdForm.Dispose();
                }
                catch (Exception ex)
                {
                }
                _createdForm = null;
            }
            for (int i = 0, loopTo = _prefixes.Count - 1; i <= loopTo; i++)
            {
                if ((_prefixes[i] ?? "") == (prefix ?? ""))
                {
                    _createdForm = AutoUIForm.Create(_settingsClients[i], _logsClients[i], _autoUiClients[i]);
                    _createdForm.Text += " RemoteApp";
                    break;
                }
            }
            return _createdForm;
        }

        public void RunRemoteApp()
        {
            if (!AvaloniaUI.IsRunning) throw new InvalidOperationException("UI is not running!");
            Dispatcher.UIThread.Invoke(() => CreateAutoUiForm().Show());
        }
    }
}